elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore/pool.rb

```diff
@@ -0,0 +1,201 @@
+              @resurrect_interval = resurrect_interval
+              @resurrectionist = start_resurrectionist
+              update_urls(urls)
+            end
+
+            def close
+              logger.debug  "Stopping resurrectionist" if logger
+              stop_resurrectionist
+              logger.debug  "Waiting for open manticore connections" if logger
+              wait_for_open_connections
+              logger.debug("Closing adapter #{@adapter}") if logger
+              @adapter.close
+            end
+
+            def wait_for_open_connections
+              until open_connections.empty?
+                logger.info "Blocked on shutdown to open connections #{@state_mutex.synchronize {@url_info}}" if logger
+                sleep 1
+              end
+            end
+
+            def open_connections
+              @state_mutex.synchronize { @url_info.values.select {|v| v[:open] > 0 } }
+            end
+
+            def alive_urls_count
+              @state_mutex.synchronize { @url_info.values.select {|v| !v[:dead] }.count }
+            end
+
+            def start_resurrectionist
+              Thread.new do
+                last_resurrect = Time.now
+                until @state_mutex.synchronize { @stopping } do
+                  if Time.now-last_resurrect >= @resurrect_interval
+                    last_resurrect = Time.now
+                    resurrect_dead!
+                  end
+                end
+              end
+            end
+
+            def resurrect_dead!
+              # Try to keep locking granularity low such that we don't affect IO...
+              @state_mutex.synchronize { @url_info.select {|url,meta| meta[:dead] } }.each do |url,meta|
+                begin
+                  @logger.info("Checking url #{url} with path #{@healthcheck_path} to see if node resurrected")
+                  perform_request_to_url(url, "GET", @healthcheck_path)
+                  # If no exception was raised it must have succeeded!
+                  logger.warn("Resurrected connection to dead ES instance at #{url}")
+                  @state_mutex.synchronize { meta[:dead] = false }
+                rescue ::Elasticsearch::Transport::Transport::HostUnreachableError => e
+                  logger.debug("Attempted to resurrect connection to dead ES instance at #{url}, got an error [#{e.class}] #{e.message}")
+                end
+              end
+            end
+
+            def stop_resurrectionist
+              @state_mutex.synchronize { @stopping = true }
+              @resurrectionist.join # Wait for thread to stop
+            end
+
+            def perform_request(method, path, params={}, body=nil)
+              with_connection do |url|
+                [url, perform_request_to_url(url, method, path, params, body)]
+              end
+            end
+
+            def perform_request_to_url(url, method, path, params={}, body=nil)
+              raise ArgumentError, "No URL specified" unless url
```

Comment:
is it worth checking this here? looking at the places where this call is made I don't see a special handling for ArgumentError.
Also, having an exception bubble up with "No URL specified" may make the user think he did something wrong?